### PR TITLE
Add container configuration

### PR DESCRIPTION
**Pull request by jthiltges**
Bring the code to python 3 compatibility, add a `--daemon` flag to run in the foreground, and make it friendlier to run in a container.

### CODE REVIEW

**Review by djw8605 — APPROVED** [review]
Looks like a lot of python 3 changes.  Looks ok though.

File: config/osg_display.conf [hunk 1]
```diff
@@ -1,8 +1,7 @@
-
 [Settings]
 timeout=3600
-logfile=/var/log/osg_display/osg_display.log
-graph_output=SVG,JPG,PNG
+#graph_output=SVG,JPG,PNG
```
Comment (edquist) [inline config/osg_display.conf:3]: Do we not want SVG anymore?

File: src/osg_display/display_graph.py [hunk 2]
```diff
@@ -63,16 +63,10 @@ def build_canvas(self, format="SVG"):
...
         else:
-            label_len = len(ylabel)
-            extra = label_len*.006
-            if label_len > 5:
-                extra += .01
-            if label_len < 5:
-                extra -= .02
-            extra = 0.0
-            ax_rect = (.05+extra, 0.06, .91-extra, .86)
+            ax_rect = (.06, 0.07, .91, .85)
```
Comment (edquist) [inline src/osg_display/display_graph.py:69]: Not that i have an opinion about them,but what is happening with these changes?  Did stuff not look good or something?

File: src/osg_display/display_graph.py [hunk 3]
```diff
@@ -185,7 +181,10 @@ def run(self, sect, mode="hourly"):
             name, tmpname = get_files(self.cp, sect, format=format)
             self.build_canvas(format=format)
             self.draw()
-            fd = open(tmpname, 'w')
+            if format in ['SVG']: # String output formats
+                fd = open(tmpname, 'w')
+            else: # Byte output formats
+                fd = open(tmpname, 'wb')
```
Comment (edquist) [inline src/osg_display/display_graph.py:187]: Just curious - is the output broken for svg with `wb` and/or non-svg without `wb`?

File: src/osg_display/gracc_datasource.py [hunk 4]
```diff
@@ -83,9 +83,8 @@ def disconnect(self):
     def connect_gracc_url(self, gracc_url):
         try:
             self.es = elasticsearch.Elasticsearch(
-                [gracc_url], timeout=300, use_ssl=True, verify_certs=True,
-                ca_certs='/etc/ssl/certs/ca-bundle.crt')
```
Comment (edquist) [inline src/osg_display/gracc_datasource.py:87]: So, just to confirm, newer ES libs work without `ca_certs` ?

File: src/osg_display/osg_display.py [hunk 5]
```diff
@@ -222,18 +223,19 @@ def main():
...
 def main():
-
-    try:
-        main_unwrapped()
-    except SystemExit:
-        raise
-    except (Exception, KeyboardInterrupt), e:
-        log.error(str(e))
-        log.exception(e)
-        raise
```
Comment (edquist) [inline src/osg_display/osg_display.py:236]: Seems you are dropping the exception logging - is this intentional?

File: src/osg_display/transfer_datasource.py [hunk 6]
```diff
@@ -124,26 +123,26 @@ def load_cached(self):
...
-        except Exception, e:
+        except Exception as e:
             log.warning("Unable to load cache; it may not exist. Error: %s" % \
                str(e))
 
     def save_cache(self):
         now = datetime.datetime.now()
         old_keys = []
-        for key in self.data.keys():
+        for key in list(self.data.keys()):
```
Comment (edquist) [inline src/osg_display/transfer_datasource.py]:
```suggestion
        for key in self.data.keys():
```

list conversion is unnecessary here

File: src/osg_display/transfer_datasource.py [hunk 7]
```diff
@@ -204,7 +203,10 @@ def query_transfers(self, starttime, endtime):
 
         response = gracc_query_transfers(self.es, transfers_raw_index, **params)
 
-        results = response.aggregations.StartTime.buckets
+        try:
+            results = response.aggregations.StartTime.buckets
+        except AttributeError:
```
Comment (edquist) [inline src/osg_display/transfer_datasource.py:208]: Interesting, does this trip over `AttributeError` sometimes?

File: src/scripts/osg_display_logrotate [hunk 8]
```diff
@@ -1,5 +0,0 @@
-/var/log/osg_display/osg_display.log {
-    missingok
-    notifempty
-    sharedscripts
-}
```
Comment (edquist) [inline src/scripts/osg_display_logrotate:5]: I guess we're not doing log rotation this way anymore?

**Review by edquist — COMMENTED** [review]
Maybe everything is fine but please see comments

File: config/osg_display.conf (hunk 1 above)
Comment (jthiltges) [reply config/osg_display.conf:3]: I updated the config file to match the running site. It's currently generating jpg and png (no svg), but only the png are being used by the site (that I could see).

File: src/osg_display/display_graph.py (hunk 2 above)
Comment (jthiltges) [reply src/osg_display/display_graph.py:69]: On the running display.osg site, the bottom edge of the graph cuts off some of the letters (note the g's and y's). With a current matplotlib, even more was cut off. I played with the numbers until it looked decent-ish.

Also, the `extra` variable gets set zero, so I took out the code to calculate it.

The graph code seemed to use lower-level matplotlib functions, and the auto-layout functions did not work. I assume that this page might eventually become a grafana dashboard, so didn't want to do a big rewrite.

File: src/osg_display/display_graph.py (hunk 3 above)
Comment (jthiltges) [reply src/osg_display/display_graph.py:187]: Yes indeed. Python 3 threw an error if the file handle didn't match the data type.

File: src/osg_display/gracc_datasource.py (hunk 4 above)
Comment (jthiltges) [reply src/osg_display/gracc_datasource.py:87]: Testing seemed fine without the `ca_certs` path. The path was specific to EL, and didn't exist on Ubuntu. (Which I ended using for the container. I don't have any strong opinions on OS, if you prefer a different base image.)

File: src/osg_display/osg_display.py (hunk 5 above)
Comment (jthiltges) [reply src/osg_display/osg_display.py:236]: Dropping the catch-all was intentional. It seemed only needed to get the exceptions into the logfile.

File: src/osg_display/transfer_datasource.py (hunk 7 above)
Comment (jthiltges) [reply src/osg_display/transfer_datasource.py:208]: In testing, I ran across cases where the StartTime attribute didn't exist in the results, causing an exception. Perhaps time ranges where there were no data points? I didn't dig into the details, and the graph output looks comparable.

File: src/scripts/osg_display_logrotate (hunk 8 above)
Comment (jthiltges) [reply src/scripts/osg_display_logrotate:5]: To make the service friendly for k8s or systemd, and centralized logging, just sending to stdout/err seemed easiest.

**Comment by edquist** [comment]
Ok!  Thanks for putting up with all my questions!  :)

I seemed to remember someone wanting svg in the past (so that the image would scale), but if it's not getting used i guess no need to generate it.

**Review by edquist — APPROVED** [review]

**Comment by jthiltges** [comment]
Thanks very much for the detailed review. I really appreciate it.

**Pull request merged**